README.md
# GluonTS Implementation of Deep Renewal Processes
[![pypi](https://img.shields.io/pypi/v/deeprenewal.svg)](https://pypi.python.org/pypi/deeprenewal)
[![travis](https://img.shields.io/travis/manujosephv/deeprenewalprocess.svg)](https://travis-ci.com/manujosephv/deeprenewalprocess)
[![documentation status](https://readthedocs.org/projects/deeprenewalprocess/badge/?version=latest)](https://deeprenewalprocess.readthedocs.io/en/latest/?badge=latest)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat-square)](https://github.com/manujosephv/deeprenewalprocess/issues)
[![HitCount](http://hits.dwyl.com/manujosephv/deeprenewalprocess.svg)](http://hits.dwyl.com/manujosephv/deeprenewalprocess)
<!-- [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Zhenye-Na/DA-RNN/blob/master/src/da_rnn.ipynb.py) -->


> Intermittent Demand Forecasting with Deep Renewal Processes
> Ali Caner Turkmen, Yuyang Wang, Tim Januschowski

## Table of Contents

- [Installation](#installation)
- [Dataset](#dataset)
    - [Download](#download)
    - [Description](#description)
- [Model](#model)
- [Usage](#usage)
    - [Train](#train)
- [Result](#result)
- [Blog](#blog)
- [References](#references)

## Installation

*Recommended Python Version*: 3.6

```
    pip install deeprenewal
```

If you are working Windows and need to use your GPU(which I recommend), you need to first install MXNet==1.6.0 version which supports GPU
[MXNet Official Installation Page](https://mxnet.apache.org/versions/1.6.0/get_started?platform=windows&language=python&processor=gpu&environ=pip&)

And if you are facing difficulties installing the GPU version, you can try(depending on the CUDA version you have)

```
pip install mxnet-cu101==1.6.0 -f https://dist.mxnet.io/python/all
```

[Relevant Github Issue](https://github.com/apache/incubator-mxnet/issues/17719)


The sources for DeepRenewal can be downloaded from the `Github repo`_.

You can either clone the public repository:

```
git clone git://github.com/manujosephv/deeprenewal
```

Once you have a copy of the source, you can install it with:
```
python setup.py install
```

## Dataset

### Download

[Retail Dataset](https://archive.ics.uci.edu/ml/datasets/online+retail)

### Description

It is a transactional data set which contains all the transactions occurring between 01/12/2010 and 09/12/2011 for a UK-based and registered non-store online retail. The company mainly sells unique all-occasion gifts. Many customers of the company are wholesalers.

Columns:

* InvoiceNo: Invoice number. Nominal, a 6-digit integral number uniquely assigned to each transaction. If this code starts with letter ‘c’, it indicates a cancellation.
* StockCode: Product (item) code. Nominal, a 5-digit integral number uniquely assigned to each distinct product.
* Description: Product (item) name. Nominal.
* Quantity: The quantities of each product (item) per transaction. Numeric.
* InvoiceDate: Invice Date and time. Numeric, the day and time when each transaction was generated.
* UnitPrice: Unit price. Numeric, Product price per unit in sterling.
* CustomerID: Customer number. Nominal, a 5-digit integral number uniquely assigned to each customer.
* Country: Country name. Nominal, the name of the country where each customer resides.

#### Preprocessing:

* Group by at StockCode, Country, InvoiceDate –> Sum of Quantity, and Mean of UnitPrice
* Filled in zeros to make timeseries continuous
* Clip lower value of Quantity to 0(removing negatives)
* Took only Time series which had length greater than 52 days.
* Train Test Split Date: 2011-11-01

#### Stats:

* No. of Timeseries: 3828. After filtering: 3671
* Quantity: Mean = 3.76, Max = 12540, Min = 0, Median = 0
* Heavily Skewed towards zero

#### Time Series Segmentation
![Segmentation](docs/imgs/1.png)

We can see that almost 98% of the timeseries in the dataset are either Intermittent or Lumpy, which is perfect for our use case.

## Model
![Architecture](docs/imgs/Deep_Renewal.png)

## Usage

### Train with CLI

```
usage: deeprenewal [-h] [--use-cuda USE_CUDA] 
                   [--datasource {retail_dataset}]
                   [--regenerate-datasource REGENERATE_DATASOURCE]
                   [--model-save-dir MODEL_SAVE_DIR]
                   [--point-forecast {median,mean}]
                   [--calculate-spec CALCULATE_SPEC] 
                   [--batch_size BATCH_SIZE]
                   [--learning-rate LEARNING_RATE] 
                   [--max-epochs MAX_EPOCHS]
                   [--number-of-batches-per-epoch NUMBER_OF_BATCHES_PER_EPOCH]
                   [--clip-gradient CLIP_GRADIENT]
                   [--weight-decay WEIGHT_DECAY]
                   [--context-length-multiplier CONTEXT_LENGTH_MULTIPLIER]
                   [--num-layers NUM_LAYERS] 
                   [--num-cells NUM_CELLS]
                   [--cell-type CELL_TYPE] 
                   [--dropout-rate DROPOUT_RATE]
                   [--use-feat-dynamic-real USE_FEAT_DYNAMIC_REAL]
                   [--use-feat-static-cat USE_FEAT_STATIC_CAT]
                   [--use-feat-static-real USE_FEAT_STATIC_REAL]
                   [--scaling SCALING]
                   [--num-parallel-samples NUM_PARALLEL_SAMPLES]
                   [--num-lags NUM_LAGS] 
                   [--forecast-type FORECAST_TYPE]

GluonTS implementation of paper 'Intermittent Demand Forecasting with Deep
Renewal Processes'

optional arguments:
  -h, --help            show this help message and exit
  --use-cuda USE_CUDA
  --datasource {retail_dataset}
  --regenerate-datasource REGENERATE_DATASOURCE
                        Whether to discard locally saved dataset and
                        regenerate from source
  --model-save-dir MODEL_SAVE_DIR
                        Folder to save models
  --point-forecast {median,mean}
                        How to estimate point forecast? Mean or Median
  --calculate-spec CALCULATE_SPEC
                        Whether to calculate SPEC. It is computationally
                        expensive and therefore False by default
  --batch_size BATCH_SIZE
  --learning-rate LEARNING_RATE
  --max-epochs MAX_EPOCHS
  --number-of-batches-per-epoch NUMBER_OF_BATCHES_PER_EPOCH
  --clip-gradient CLIP_GRADIENT
  --weight-decay WEIGHT_DECAY
  --context-length-multiplier CONTEXT_LENGTH_MULTIPLIER
                        If context multipler is 2, context available to hte
                        RNN is 2*prediction length
  --num-layers NUM_LAYERS
  --num-cells NUM_CELLS
  --cell-type CELL_TYPE
  --dropout-rate DROPOUT_RATE
  --use-feat-dynamic-real USE_FEAT_DYNAMIC_REAL
  --use-feat-static-cat USE_FEAT_STATIC_CAT
  --use-feat-static-real USE_FEAT_STATIC_REAL
  --scaling SCALING     Whether to scale targets or not
  --num-parallel-samples NUM_PARALLEL_SAMPLES
  --num-lags NUM_LAGS   Number of lags to be included as feature
  --forecast-type FORECAST_TYPE
                        Defines how the forecast is decoded. For details look
                        at the documentation
```

An example of training process is as follows:

```
python3 deeprenewal --datasource retail_dataset --lr 0.001 --epochs 50
```
### Train with Jupyter Notebook

Check out the examples folder for notebooks

## Result

| Method             | QuantileLoss[0.25] | QuantileLoss[0.5] | QuantileLoss[0.75] | mean_wQuantileLoss |
|--------------------|--------------------|-------------------|--------------------|--------------------|
| Croston            | 664896.9323        | 791880.3858       | 918863.8392        | 1.034257626        |
| SBA                | 623338.1011        | 776084.5519       | 928831.0028        | 1.013627034        |
| SBJ                | 627880.7754        | 779758.6188       | 931636.4622        | 1.018425652        |
| ARIMA              | 598779.2977        | 784662.7412       | 957980.814         | 1.019360367        |
| ETS                | 622502.7789        | 796128.4          | 957808.4087        | 1.03460523         |
| DeepAR             | **378217.1822**        | **679862.7643**       | **808336.3482**        | **0.812561813**        |
| NPTS               | 380956             | 725255            | 935102.5           | 0.88870495         |
| DeepRenewal Flat   | 383524.4007        | 764167.8638       | 1047169.894        | 0.955553796        |
| DeepRenewal Exact  | 382825.5           | 765640            | 1141210.5          | 0.99683189         |
| DeepRenewal Hybrid | 389981.2253        | 761474.4966       | 1069187.032        | 0.96677762         |

## Blog

For a more detailed account of the implementation and the experiments please visit the blog:


## References

[1] Ali Caner Turkmen, Yuyang Wang, Tim Januschowski. [*"Intermittent Demand Forecasting with Deep Renewal Processes"*](https://arxiv.org/pdf/1911.10416.pdf). arXiv:1911.10416 [cs.LG] (2019)
[2] Alexander Alexandrov, Konstantinos Benidis, Michael Bohlke-Schneider, Valentin Flunkert, Jan Gasthaus, Tim Januschowski, Danielle C. Maddix, Syama Rangapuram, David Salinas, Jasper Schulz, Lorenzo Stella, Ali Caner Türkmen, Yuyang Wang;. [*"GluonTS: Probabilistic and Neural Time Series Modeling in Python"*](https://www.jmlr.org/papers/v21/19-820.html). (2020).

HISTORY.rst
=======
History
=======

0.1.2 (2020-07-24)
------------------

* Minor changes in documentation.

0.1.0 (2020-07-24)
------------------

* First release on PyPI.


deeprenewal/cli.py
"""Console script for deeprenewal."""
import random
import os
import numpy as np
import mxnet as mx
from pathlib import Path
from tqdm.autonotebook import tqdm

def seed_everything():
    random.seed(42)
    np.random.seed(42)
    mx.random.seed(42)


seed_everything()

from gluonts.trainer import Trainer
from gluonts.evaluation.backtest import make_evaluation_predictions
from argparse import ArgumentParser
import sys
import ast
from deeprenewal import get_dataset, DeepRenewalEstimator, IntermittentEvaluator


def evaluate_forecast(args):
    is_gpu = mx.context.num_gpus() > 0
    dataset = get_dataset(args.datasource, regenerate=args.regenerate_datasource)
    prediction_length = dataset.metadata.prediction_length
    freq = dataset.metadata.freq
    cardinality = (
        ast.literal_eval(dataset.metadata.feat_static_cat[0].cardinality)
        if args.use_feat_static_cat
        else None
    )
    train_ds = dataset.train
    test_ds = dataset.test
    trainer = Trainer(
        ctx=mx.context.gpu() if is_gpu & args.use_cuda else mx.context.cpu(),
        batch_size=args.batch_size,
        learning_rate=args.learning_rate,
        epochs=args.max_epochs,
        num_batches_per_epoch=args.number_of_batches_per_epoch,
        clip_gradient=args.clip_gradient,
        weight_decay=args.weight_decay,
        hybridize=True,
    )  # hybridize false for development

    estimator = DeepRenewalEstimator(
        prediction_length=prediction_length,
        context_length=prediction_length * args.context_length_multiplier,
        num_layers=args.num_layers,
        num_cells=args.num_cells,
        cell_type=args.cell_type,
        dropout_rate=args.dropout_rate,
        scaling=True,
        lags_seq=np.arange(1, args.num_lags + 1).tolist(),
        freq=freq,
        use_feat_dynamic_real=args.use_feat_dynamic_real,
        use_feat_static_cat=args.use_feat_static_cat,
        use_feat_static_real=args.use_feat_static_real,
        cardinality=cardinality if args.use_feat_static_cat else None,
        trainer=trainer,
    )
    predictor = estimator.train(train_ds, test_ds)
    os.makedirs(args.model_save_dir, exist_ok=True)
    os.makedirs(Path(args.model_save_dir)/"deep_renewal", exist_ok=True)
    predictor.serialize(Path(args.model_save_dir)/"deep_renewal")
    print("Generating DeepRenewal forecasts.......")
    # predictor = Predictor.deserialize(Path(args.model_save_dir)/"deeprenewal")
    deep_renewal_flat_forecast_it, ts_it = make_evaluation_predictions(
        dataset=test_ds, predictor=predictor, num_samples=100
    )
    tss = list(tqdm(ts_it, total=len(test_ds)))
    deep_renewal_flat_forecasts = list(tqdm(deep_renewal_flat_forecast_it, total=len(test_ds)))
    evaluator = IntermittentEvaluator(quantiles=[0.25,0.5,0.75], median=args.point_forecast=="median", calculate_spec=args.calculate_spec)
    deep_renewal_flat_agg_metrics, deep_renewal_flat_item_metrics = evaluator(
        iter(tss), iter(deep_renewal_flat_forecasts), num_series=len(test_ds)
    )
    print(deep_renewal_flat_agg_metrics)



def main():
    """Console script for deeprenewal."""
    parser = ArgumentParser(description="GluonTS implementation of paper 'Intermittent Demand Forecasting with Deep Renewal Processes'")
    # PROGRAM level args
    parser.add_argument("--use-cuda", type=bool, default=True)
    parser.add_argument("--datasource", type=str, default="retail_dataset", choices=["retail_dataset"])
    parser.add_argument("--regenerate-datasource", help="Whether to discard locally saved dataset and regenerate from source", type=bool, default=False)
    parser.add_argument("--model-save-dir", help="Folder to save models", type=str, default="saved_models")
    parser.add_argument("--point-forecast", help="How to estimate point forecast? Mean or Median", type=str, default="mean", choices=['median', 'mean'])
    parser.add_argument("--calculate-spec", help="Whether to calculate SPEC. It is computationally expensive and therefore False by default", type=bool, default=False)

    # Trainer specific args
    parser.add_argument("--batch_size", type=int, default=32)
    parser.add_argument("--learning-rate", type=float, default=1e-2)
    parser.add_argument("--max-epochs", type=int, default=20)
    parser.add_argument("--number-of-batches-per-epoch", type=int, default=100)
    parser.add_argument("--clip-gradient", type=float, default=5.170127652392614)
    parser.add_argument("--weight-decay", type=float, default=0.01)

    # Model specific args
    parser.add_argument("--context-length-multiplier", help="If context multipler is 2, context available to hte RNN is 2*prediction length", type=int, default=2)
    parser.add_argument("--num-layers", type=int, default=2)
    parser.add_argument("--num-cells", type=int, default=64)
    parser.add_argument("--cell-type", type=str, default="lstm")
    # p% are dropped and set to zero
    parser.add_argument("--dropout-rate", type=float, default=0.3)
    parser.add_argument("--use-feat-dynamic-real", type=bool, default=False)
    parser.add_argument("--use-feat-static-cat", type=bool, default=False)
    parser.add_argument("--use-feat-static-real", type=bool, default=False)
    parser.add_argument("--scaling", help="Whether to scale targets or not", type=bool, default=True)
    parser.add_argument("--num-parallel-samples", type=int, default=100)
    parser.add_argument("--num-lags", help="Number of lags to be included as feature", type=int, default=1)
    parser.add_argument("--forecast-type", help="Defines how the forecast is decoded. For details look at the documentation", type=str, default="hybrid")
    args = parser.parse_args()

    print("Arguments: " + str(args._))
    evaluate_forecast(args)
    return 0


if __name__ == "__main__":
    sys.exit(main())  # pragma: no cover
